common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -1220,6 +1260,7 @@ public void onData(String streamId, ByteBuffer buf) throws IOException {
           try {
             if (deferredBufs != null && !deferredBufs.isEmpty()) {
               writeDeferredBufs();
+              mergeManager.pushMergeMetrics.deferredBlocks.mark();
```

Comment:
This is when the deferred block is getting written out.
Instead, we should be incrementing it when the deferred block is getting added.
Move it to where we increment `deferredBlockBytes` below ?
This will keep the metrics in sync - else we can have a large `deferredBlockBytes`, but very low `deferredBlockBytes`.

An example of this is in the test case [here](https://github.com/apache/spark/pull/37638/files#diff-955b4eb8f495555d5e188ed93a4acf5393e1d9c7855dffddedb5e9201eebdef2R193) - the deferred blocks should be `1`m but is `0`

Reply:
Sure. In addition to the proposal here, to keep the `deferredBlockBytes` and `deferredBlocks ` fully in sync, I also went ahead and decrease the `deferredBlocks` whenever there is a `deferredBlockBytes.dec()` and accommodate the UTs, please take a look.